clis/teliod/src/core_api.rs

```diff
@@ -0,0 +1,338 @@
+    traffic_routing_supported: bool,
+}
+
+pub async fn init_with_api(
+    auth_token: &str,
+    interface_name: &str,
+) -> Result<DeviceIdentity, Error> {
+    let mut identity_path = dirs::data_local_dir().ok_or(Error::NoDataLocalDir)?;
+    identity_path.push("teliod");
+    if !identity_path.exists() {
+        let _ = create_dir_all(&identity_path);
+    }
+    identity_path.push(&format!("{interface_name}.json"));
+
+    let device_identity = match DeviceIdentity::from_file(&identity_path) {
+        Some(identity) => identity,
+        None => {
+            let private_key = DeviceIdentity::generate_sk();
+            let hw_identifier = DeviceIdentity::generate_hw_identifier(private_key);
+
+            let machine_identifier =
+                match fetch_identifier_with_exp_backoff(auth_token, private_key.public()).await {
+                    Ok(id) => id,
+                    Err(e) => match e {
+                        Error::DeviceNotFound => {
+                            debug!("Unable to load identifier due to {e}. Registering ...");
+                            register_machine_with_exp_backoff(
+                                &hw_identifier,
+                                private_key.public(),
+                                auth_token,
+                            )
+                            .await?
+                        }
+                        _ => return Err(e),
+                    },
+                };
+
+            DeviceIdentity {
+                private_key,
+                hw_identifier,
+                machine_identifier,
+            }
+        }
+    };
+
+    let mut backoff = ExponentialBackoffBuilder::new()
+        .with_max_elapsed_time(Some(Duration::from_secs(300)))
+        .build();
+
+    loop {
+        let status = update_machine(&device_identity, auth_token).await?;
+        match status {
+            StatusCode::OK => break,
+            StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED | StatusCode::BAD_REQUEST => {
+                // exit daemon
+                return Err(Error::UpdateMachineError(status));
+            }
+            _ => {
+                // Retry with exp back-off
+                let _ = wait_with_backoff_delay(
+                    &mut backoff,
+                    "update machine",
+                    Error::UpdateMachineError(status),
+                )
+                .await?;
+            }
+        }
+    }
+
+    let mut options = OpenOptions::new();
+    // User has read/write but others have none
+    options.mode(0o600);
+    let mut file = options.create(true).write(true).open(identity_path)?;
+    serde_json::to_writer(&mut file, &device_identity)?;
+    Ok(device_identity)
+}
+
+async fn fetch_identifier_with_exp_backoff(
+    auth_token: &str,
+    public_key: PublicKey,
+) -> Result<MachineIdentifier, Error> {
+    let mut backoff = ExponentialBackoffBuilder::new()
+        .with_max_elapsed_time(Some(Duration::from_secs(300)))
```

Comment:
maybe we could define `300` as a global static?

Reply:
Made it into a file local const